crypto/signature.go

```diff
@@ -0,0 +1,196 @@
+// Copyright 2025 Democratized Data Foundation
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package crypto
+
+import (
+	"crypto/ed25519"
+	"crypto/sha256"
+
+	"github.com/decred/dcrd/dcrec/secp256k1/v4"
+	"github.com/decred/dcrd/dcrec/secp256k1/v4/ecdsa"
+)
+
+// Sign signs the provided message using the appropriate signature algorithm based on the key type.
+//
+// This is a generic function that accepts different types of private keys and
+// automatically uses the correct signing algorithm based on the key type.
+//
+// For ECDSA (secp256k1):
+// - Returns signature in DER format
+// - Accepts private key as *secp256k1.PrivateKey
+//
+// For Ed25519:
+// - Returns standard Ed25519 signature
+// - Accepts private key as ed25519.PrivateKey
+//
+// Parameters:
+//   - privKey: The private key to sign with
+//   - message: The message to sign
+//
+// Returns:
+//   - []byte: The signature in the format appropriate for the key type
+//   - error: Any error encountered during signing, including invalid key types
+func Sign[T ed25519.PrivateKey | *secp256k1.PrivateKey](
+	privKey T,
+	message []byte,
+) ([]byte, error) {
+	switch k := any(privKey).(type) {
+	case *secp256k1.PrivateKey:
+		return SignECDSA256K(k, message)
+	case ed25519.PrivateKey:
+		return SignEd25519(k, message)
+	default:
+		// This should never happen due to type constraints on T
+		return nil, ErrUnsupportedPrivKeyType
```

Comment:
question: Did the Go compiler require this (or a return outside the switch)?

Reply:
Ah sorry, some languages allow you to write (roughly) the below, and the compiler knows that all the switch handles all the possible types, removing the need for `default`.  I tested though and it looks like Go doesn't yet support this:
```
func Sign[T ed25519.PrivateKey | *secp256k1.PrivateKey](
	privKey T,
	message []byte,
) ([]byte, error) {
	switch k := any(privKey).(type) {
	case *secp256k1.PrivateKey:
		return SignECDSA256K(k, message)
	case ed25519.PrivateKey:
		return SignEd25519(k, message)
	}
}
```